Zend/zend_compile.c

```diff
@@ -2296,7 +2305,7 @@ static void zend_emit_return_type_check(
 			}
 		}
 
-		if (expr && expr->op_type == IS_CONST) {
+		if (expr && (expr->op_type == IS_CONST || ZEND_TYPE_PURE_MASK(type) == MAY_BE_ANY)) {
```

Comment:
It seems like reading something that isn't IS_CONST (i.e. IS_CV, IS_TMP_VAR, IS_VAR) as if it were actually IS_CONST expr->u.constant would read invalid memory?) e.g. `return $cv;` or `return $a + $b;`

Splitting this up would be clearer

```suggestion
		if (expr && ZEND_TYPE_PURE_MASK(type) == MAY_BE_ANY) {
			/* we don't need run-time check for mixed return type */
			return;
		}
		if (expr && expr->op_type == IS_CONST) {
```

Reply:
Thanks for the suggestion! It makes sense for me :)